### ISSUE TITLE
Error: Cannot use soft decisions during evaluation.

### ISSUE THREAD

**Issue by oliverguhr** [pbelcak/fastfeedforward#2]
Hello,
thanks for publishing the code along with your paper. While reading your paper, I tried to run the demo notebook from this repository. I modified it slightly for colab to use the pip package instead of the local repository. However, I run into this issue:

```python
ValueError                                Traceback (most recent call last)
<ipython-input-6-809883d6966a> in <cell line: 10>()
     32         testing_accuracies = []
     33         for batch_images, batch_labels in tqdm(dataloader_testing):
---> 34                 output = model(batch_images.view(-1, 784))
     35                 loss = criterion(output, batch_labels)
     36                 accuracy = (output.argmax(dim=1) == batch_labels).detach().float().mean()

1 frames
/usr/local/lib/python3.10/dist-packages/fastfeedforward/fff.py in forward(self, x, return_entropies, use_hard_decisions)
    332                                 raise ValueError("Cannot return entropies during evaluation.")
    333                         if not use_hard_decisions:
--> 334                                 raise ValueError("Cannot use soft decisions during evaluation.")
    335                         return self.eval_forward(x)
    336 

ValueError: Cannot use soft decisions during evaluation.
```
Here is the notebook to reproduce the error:

https://colab.research.google.com/drive/1apDlwbJPGMMoNodTOUx5m8Klnp8zE7T-?usp=sharing

**Comment by smpurkis** [comment]
I'm getting the same error. 

I'm getting a "Need access error" for the above Colab notebook.

So here is a short script to reproduce.

Versions used:
- fastfeedforward - 0.2.0
- torch - 2.0.1

```
import torch.nn as nn
from fastfeedforward import FFF
import math
import torch

print(torch.__version__)

input_width = 512
hidden_width = 512
output_width = 10
dropout = 0.5
depth = 5
leaf_width = math.ceil(hidden_width / 2**depth)
region_leak = 0.2  # your choice of the region leak probability (0 - 0.5) to encourage generalisation in very deep FFFs


my_fff = FFF(
    input_width,
    leaf_width,
    output_width,
    depth,
    activation=nn.ReLU(),
    dropout=dropout,
    region_leak=region_leak,
)


class FFFNeuralNetwork(nn.Module):
    def __init__(self):
        super(FFFNeuralNetwork, self).__init__()
        self.flatten = nn.Flatten()
        self.linear_relu_stack = nn.Sequential(
            nn.Linear(28 * 28, input_width), nn.ReLU(), nn.Dropout(dropout), my_fff
        )

    def forward(self, x):
        x = self.flatten(x)
        logits = self.linear_relu_stack(x)
        return logits


model = FFFNeuralNetwork()

test_tensor = torch.randn(1, 1, 28, 28)
model.eval()

output = model(test_tensor)

print(output)
```

The output:
```
2.0.1
Traceback (most recent call last):
  File "/home/ubuntu/projects/feedfast-pg/small.py", line 47, in <module>
    output = model(test_tensor)
  File "/home/ubuntu/projects/feedfast-pg/.venv/lib/python3.10/site-packages/torch/nn/modules/module.py", line 1501, in _call_impl
    return forward_call(*args, **kwargs)
  File "/home/ubuntu/projects/feedfast-pg/small.py", line 38, in forward
    logits = self.linear_relu_stack(x)
  File "/home/ubuntu/projects/feedfast-pg/.venv/lib/python3.10/site-packages/torch/nn/modules/module.py", line 1501, in _call_impl
    return forward_call(*args, **kwargs)
  File "/home/ubuntu/projects/feedfast-pg/.venv/lib/python3.10/site-packages/torch/nn/modules/container.py", line 217, in forward
    input = module(input)
  File "/home/ubuntu/projects/feedfast-pg/.venv/lib/python3.10/site-packages/torch/nn/modules/module.py", line 1501, in _call_impl
    return forward_call(*args, **kwargs)
  File "/home/ubuntu/projects/feedfast-pg/.venv/lib/python3.10/site-packages/fastfeedforward/fff.py", line 334, in forward
    raise ValueError("Cannot use soft decisions during evaluation.")
ValueError: Cannot use soft decisions during evaluation.
```

**Comment by smpurkis** [comment]
I have found a workaround, but I still think this is a bug.

In my example above the model class code:
```
my_fff = FFF(
    input_width,
    leaf_width,
    output_width,
    depth,
    activation=nn.ReLU(),
    dropout=dropout,
    region_leak=region_leak,
)


class FFFNeuralNetwork(nn.Module):
    def __init__(self):
        super(FFFNeuralNetwork, self).__init__()
        self.flatten = nn.Flatten()
        self.linear_relu_stack = nn.Sequential(
            nn.Linear(28 * 28, input_width), nn.ReLU(), nn.Dropout(dropout), my_fff
        )

    def forward(self, x):
        x = self.flatten(x)
        logits = self.linear_relu_stack(x)
        return logits
``` 

 becomes
 ```
 class FFFNeuralNetwork(nn.Module):
    def __init__(self):
        super(FFFNeuralNetwork, self).__init__()
        self.flatten = nn.Flatten()
        self.linear_relu_stack = nn.Sequential(
            nn.Linear(28 * 28, input_width), nn.ReLU(), nn.Dropout(dropout)
        )
        self.fff = FFF(
            input_width,
            leaf_width,
            output_width,
            depth,
            activation=nn.ReLU(),
            dropout=dropout,
            region_leak=region_leak,
        )

    def forward(self, x):
        x = self.flatten(x)
        logits = self.linear_relu_stack(x)
        logits = self.fff(
            logits, use_hard_decisions=True if not self.fff.training else None
        )
        return logits
```

i.e. I make the FFF class its own parameter in the class and call it explicitly in the forward method, with the keyword parameter `use_hard_decisions=True if not self.fff.training else None`

Then running again the output becomes
```
2.0.1
tensor([[-0.1805, -0.2155,  0.1804,  0.0836,  0.0451,  0.1837, -0.1774, -0.0666,
         -0.0719,  0.0760]], grad_fn=<ViewBackward0>)
```

**Comment by pbelcak** [comment]
Hi @oliverguhr @smpurkis ,

Great issue, thanks for pointing it out! I believe this was fixed in `v0.2.1`, which went live on PyPI today.
Could you please `pip install fastfeedforward --upgrade` and confirm?

**Comment by oliverguhr** [comment]
Thanks a lot @pbelcak! Version 0.2.1 fixed it, so I am closing this issue.
